File: 05-memory-management/mmap.c
#include <unistd.h>
#include <sys/mman.h>
#include <stdio.h>
#include <stdlib.h>
#include <err.h>

#define BUFFER_SIZE 1000
#define ALLOC_SIZE (100*1024*1024)

static char command[BUFFER_SIZE];

int main(void)
{
	pid_t pid;

	pid = getpid();
	snprintf(command, BUFFER_SIZE, "cat /proc/%d/maps", pid);

	puts("*** memory map before memory allocation ***");
	fflush(stdout);
	system(command);

	void *new_memory;
	new_memory = mmap(NULL, ALLOC_SIZE, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
	if (new_memory == (void *) -1)
		err(EXIT_FAILURE, "mmap() failed");

	puts("");
	printf("*** succeeded to allocate memory: address = %p; size = 0x%x ***\n",
		new_memory, ALLOC_SIZE);
	puts("");
	
	puts("*** memory map after memory allocation ***");
	fflush(stdout);
	system(command);

	if (munmap(new_memory, ALLOC_SIZE) == -1)
		err(EXIT_FAILURE, "munmap() failed");
	exit(EXIT_SUCCESS);
}

File: 05-memory-management/misc/mmap.go
package main

import (
	"fmt"
	"log"
	"os"
	"os/exec"
	"strconv"
	"syscall"
)

const (
	ALLOC_SIZE = 1024 * 1024 * 1024
)

func main() {
	pid := os.Getpid()
	fmt.Println("*** 新規メモリ領域獲得前のメモリマップ ***")
	command := exec.Command("cat", "/proc/"+strconv.Itoa(pid)+"/maps")
	command.Stdout = os.Stdout
	err := command.Run()
	if err != nil {
		log.Fatal("catの実行に失敗しました")
	}

	// mmap()システムコールの呼び出しによって1GBのメモリ領域を獲得
	data, err := syscall.Mmap(-1, 0, ALLOC_SIZE, syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_ANON|syscall.MAP_PRIVATE)
	if err != nil {
		log.Fatal("mmap()に失敗しました")
	}

	fmt.Println("")
	fmt.Printf("*** 新規メモリ領域: アドレス = 0x%p, サイズ = 0x%x ***\n",
		&data[0], ALLOC_SIZE)
	fmt.Println("")

	fmt.Println("*** 新規メモリ領域獲得後のメモリマップ ***")
	command = exec.Command("cat", "/proc/"+strconv.Itoa(pid)+"/maps")
	command.Stdout = os.Stdout
	err = command.Run()
	if err != nil {
		log.Fatal("catの実行に失敗しました")
	}
}
